### PR TITLE
fix: `ERR_PNPM_TARBALL_EXTRACT` when the URL's hash contains a slash

### PR DESCRIPTION
**Pull request by IceOfSummer**
Before determine the URL type, encode the hash if it contains a slash. 

Fix: #7697

### CODE REVIEW

**Comment by IceOfSummer** [comment]
Hello! There is an Eslint error in upstream/main:
```log
pnpm/workspace/state/src/createWorkspaceState.ts
  13:20  error  Require 'compare' argument  @typescript-eslint/require-array-sort-compare
```

I will rebase my branch after this error is resolved. Or should I resolve it myself in this PR?

File: resolving/tarball-resolver/src/index.ts [hunk 1]
```diff
@@ -3,7 +3,7 @@ import { type PkgResolutionId, type ResolveResult } from '@pnpm/resolver-base'
 export async function resolveFromTarball (
   wantedDependency: { pref: string }
 ): Promise<ResolveResult | null> {
-  if (!wantedDependency.pref.startsWith('http:') && !wantedDependency.pref.startsWith('https:')) {
+  if (!(wantedDependency.pref.startsWith('http:') || wantedDependency.pref.startsWith('https:'))) {
```
Comment (zkochan) [inline resolving/tarball-resolver/src/index.ts]: why was this changed? it is unrelated to the fix

File: pnpm/test/install/misc.ts [hunk 2]
```diff
@@ -561,3 +561,15 @@ test('do not hang on circular peer dependencies', () => {
...
   expect(fs.existsSync(path.join(tempDir, WANTED_LOCKFILE))).toBeTruthy()
 })
+
+// Covers https://github.com/pnpm/pnpm/issues/7697
+test('install success even though the url\'s hash contains slash', async () => {
+  prepare()
+  const settings = ['--fetch-retries=0']
+  const result = execPnpmSync([
+    'add',
+    'https://github.com/actions/toolkit.git#core/1.10',
```
Comment (zkochan) [inline pnpm/test/install/misc.ts]:
```suggestion
    'https://github.com/pnpm-e2e/simple-pkg.git#branch/with-slash',
```

**Review by zkochan — CHANGES_REQUESTED** [review]

File: resolving/tarball-resolver/src/index.ts (hunk 1 above)
Comment (vudsen) [reply resolving/tarball-resolver/src/index.ts]: Sorry, this is my own hobby. This writing is better for my understanding...

**Review by zkochan — APPROVED** [review]

**Pull request merged**